app/controllers/carto/api/visualizations_controller.rb

```
@@ -99,6 +111,58 @@ def is_liked
         })
       end
 
+      def add_like
+        current_viewer_id = current_viewer.id
+
+        @visualization.add_like_from(current_viewer_id)
+
+        if current_viewer_id != @visualization.user.id
+          protocol = request.protocol.sub('://', '')
+          vis_url =
+            Carto::StaticMapsURLHelper.new.url_for_static_map_with_visualization(@visualization, protocol, 600, 300)
+          send_like_email(@visualization, current_viewer, vis_url)
+        end
+
+        event_properties = { user_id: current_viewer_id, visualization_id: @visualization.id, action: 'like' }
+        Carto::Tracking::Events::LikedMap.new(current_viewer_id, event_properties).report
+
+        render_jsonp(
+          id: @visualization.id,
+          likes: @visualization.likes.count,
+          liked: @visualization.liked_by?(current_viewer_id)
+        )
+      rescue Carto::Visualization::AlreadyLikedError
+        render(text: "You've already liked this visualization", status: 400)
+      rescue KeyError => exception
+        render(text: exception.message, status: 403)
```

Yep, please :-) Old controllers had tons of silenced errors that made some bugs be unnoticed for years. IMHO, there are not many errors that should just remain hidden.